File: Working/Baek_1182.cpp
/*
    @ Baek 1182
    @ Prob. https://www.acmicpc.net/problem/1182
      Ref.
    @ Algo: Backtracking
    @ Start day: 19. 12. 17
    @ End day:

*/


#include<iostream>
#include<algorithm>
#include<vector>

using namespace std;

int answer = 0;
int N;
int result;

void go(vector<int> numVec, int index, int sum) {

	if (N == index) {
		if (sum == result) {
			answer += 1;
		}
		return;
	}

	go(numVec, index + 1, sum + numVec[index]);
	go(numVec, index + 1, sum);
}

int main() {
	cin >> N>> result;
	vector<int> numVec(N);

	for (int i = 0; i < N; i++) {
		cin >> numVec[i];
	}

	go(numVec, 0, 0);
	if (result == 0) answer -= 1;
	cout << answer << endl;

	return 0;
}

File: DP/Baek_1182.py
"""
    @ 1182. 부분수열의 합
    @ Prob. https://www.acmicpc.net/problem/1182
     Ref.
    @ Algo: DP
    @ Start day: 20. 01. 31.
    @ End day: 20. 01. 31.
"""

def solution(idx, total_sum):
    global result

    if idx >= N:
        return
    total_sum += arr[idx]

    if total_sum == S:
        result += 1

    solution(idx+1, total_sum-arr[idx])
    solution(idx+1, total_sum)


result = 0
N, S = map(int, input().split())
arr = list(map(int, input().split()))

solution(0, 0)
print(result)

"""
5 0
-7 -3 -2 5 8
>
1
"""